build: multiprocess link issues on fedora aarch64

master @ ad09b762757a293ff970113c15bd9dee3057e556
```bash
make -C depends/ -j5 NO_QT=1 NO_WALLET=1 NO_ZMQ=1 NO_NATPMP=1 NO_UPNP=1 MULTIPROCESS=1
./autogen.sh
CONFIG_SITE=/home/fedora/bitcoin/depends/aarch64-unknown-linux-gnu/share/config.site ./configure
make
...
  CXXLD    bitcoin-node
  CXXLD    test/test_bitcoin
/usr/bin/ld: /home/fedora/bitcoin/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o): relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `_ZN2kj1_5Debug11minSeverityE' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /home/fedora/bitcoin/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o)(.text+0xa34): unresolvable R_AARCH64_ADR_PREL_PG_HI21 relocation against symbol `_ZN2kj1_5Debug11minSeverityE'
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

> It happens with clang-15 which is the default on Fedora 37.

No? clang is not default for Fedora. g++ is picked, at least on my system.

The bug doesn't happen on Lunar. On Fedora it happens with both g++ and clang++:


```
# make V=1
Making all in src
make[1]: Entering directory '/bitcoin-core/src'
make[2]: Entering directory '/bitcoin-core/src'
make[3]: Entering directory '/bitcoin-core'
make[3]: Leaving directory '/bitcoin-core'
/bin/sh ../libtool  --tag=CXX --preserve-dup-deps  --mode=link /usr/bin/ccache g++ -std=c++17 -fdebug-prefix-map=/bitcoin-core=. -fstack-reuse=none -Wstack-protector -fstack-protector-all -fstack-clash-protection       -fno-extended-identifiers -fPIE -pipe -std=c++17 -O2    -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -pie      -pthread -lpthread -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib  -o bitcoin-node bitcoin_node-bitcoind.o  init/bitcoin_node-bitcoin-node.o libbitcoin_node.a  libbitcoin_common.a libbitcoin_util.a libunivalue.la  libbitcoin_consensus.a crypto/libbitcoin_crypto_base.la    crypto/libbitcoin_crypto_arm_shani.la leveldb/libleveldb.la crc32c/libcrc32c.la   leveldb/libmemenv.la secp256k1/libsecp256k1.la    -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -levent_pthreads -levent  -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -levent    libbitcoin_ipc.a -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -lmultiprocess -lcapnp-rpc -lcapnp -lkj-async -lkj -pthread -lpthread  
libtool: link: /usr/bin/ccache g++ -std=c++17 -fdebug-prefix-map=/bitcoin-core=. -fstack-reuse=none -Wstack-protector -fstack-protector-all -fstack-clash-protection -fno-extended-identifiers -fPIE -pipe -std=c++17 -O2 -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-z -Wl,separate-code -pie -o bitcoin-node bitcoin_node-bitcoind.o init/bitcoin_node-bitcoin-node.o  -lpthread -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib libbitcoin_node.a libbitcoin_common.a libbitcoin_util.a ./.libs/libunivalue.a libbitcoin_consensus.a crypto/.libs/libbitcoin_crypto_base.a crypto/.libs/libbitcoin_crypto_arm_shani.a leveldb/.libs/libleveldb.a crc32c/.libs/libcrc32c.a leveldb/.libs/libmemenv.a secp256k1/.libs/libsecp256k1.a -levent_pthreads -levent -levent libbitcoin_ipc.a -lmultiprocess /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libcapnp-rpc.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libcapnp.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libkj-async.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libkj.so -lpthread -pthread -Wl,-rpath -Wl,/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -Wl,-rpath -Wl,/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib
/usr/bin/ld: /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o): relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `_ZN2kj1_5Debug11minSeverityE' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o)(.text+0xa34): unresolvable R_AARCH64_ADR_PREL_PG_HI21 relocation against symbol `_ZN2kj1_5Debug11minSeverityE'
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:7114: bitcoin-node] Error 1
make[2]: Leaving directory '/bitcoin-core/src'
make[1]: *** [Makefile:19201: all-recursive] Error 1
make[1]: Leaving directory '/bitcoin-core/src'
make: *** [Makefile:817: all-recursive] Error 1
```

```
# make V=1
Making all in src
make[1]: Entering directory '/bitcoin-core/src'
make[2]: Entering directory '/bitcoin-core/src'
make[3]: Entering directory '/bitcoin-core'
make[3]: Leaving directory '/bitcoin-core'
/bin/sh ../libtool  --tag=CXX --preserve-dup-deps  --mode=link /usr/bin/ccache clang++ -std=c++17 -fdebug-prefix-map=/bitcoin-core=. -Wstack-protector -fstack-protector-all        -fPIE -pipe -std=c++17 -O2    -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -pie      -pthread -lpthread -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib  -o bitcoin-node bitcoin_node-bitcoind.o  init/bitcoin_node-bitcoin-node.o libbitcoin_node.a  libbitcoin_common.a libbitcoin_util.a libunivalue.la  libbitcoin_consensus.a crypto/libbitcoin_crypto_base.la    crypto/libbitcoin_crypto_arm_shani.la leveldb/libleveldb.la crc32c/libcrc32c.la  crc32c/libcrc32c_arm_crc.la leveldb/libmemenv.la secp256k1/libsecp256k1.la    -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -levent_pthreads -levent  -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -levent    libbitcoin_ipc.a -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -lmultiprocess -lcapnp-rpc -lcapnp -lkj-async -lkj -pthread -lpthread  
libtool: link: /usr/bin/ccache clang++ -std=c++17 -fdebug-prefix-map=/bitcoin-core=. -Wstack-protector -fstack-protector-all -fPIE -pipe -std=c++17 -O2 -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-z -Wl,separate-code -pie -o bitcoin-node bitcoin_node-bitcoind.o init/bitcoin_node-bitcoin-node.o  -lpthread -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib libbitcoin_node.a libbitcoin_common.a libbitcoin_util.a ./.libs/libunivalue.a libbitcoin_consensus.a crypto/.libs/libbitcoin_crypto_base.a crypto/.libs/libbitcoin_crypto_arm_shani.a leveldb/.libs/libleveldb.a crc32c/.libs/libcrc32c.a crc32c/.libs/libcrc32c_arm_crc.a leveldb/.libs/libmemenv.a secp256k1/.libs/libsecp256k1.a -levent_pthreads -levent -levent libbitcoin_ipc.a -lmultiprocess /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libcapnp-rpc.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libcapnp.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libkj-async.so /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libkj.so -lpthread -pthread -Wl,-rpath -Wl,/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib -Wl,-rpath -Wl,/bitcoin-core/depends/aarch64-unknown-linux-gnu/lib
/usr/bin/ld: /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o): relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `_ZN2kj1_5Debug11minSeverityE' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /bitcoin-core/depends/aarch64-unknown-linux-gnu/lib/libmultiprocess.a(proxy.cpp.o)(.text+0x10): unresolvable R_AARCH64_ADR_PREL_PG_HI21 relocation against symbol `_ZN2kj1_5Debug11minSeverityE'
/usr/bin/ld: final link failed: bad value
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:7114: bitcoin-node] Error 1
make[2]: Leaving directory '/bitcoin-core/src'
make[1]: *** [Makefile:19201: all-recursive] Error 1
make[1]: Leaving directory '/bitcoin-core/src'
make: *** [Makefile:817: all-recursive] Error 1

--- a/depends/packages/capnp.mk
+++ b/depends/packages/capnp.mk
@@ -5,10 +5,15 @@ $(package)_download_file=$(native_$(package)_download_file)
 $(package)_file_name=$(native_$(package)_file_name)
 $(package)_sha256_hash=$(native_$(package)_sha256_hash)
 
+# Hardcode library install path to "lib" to match the PKG_CONFIG_PATH
+# setting in depends/config.site.in, which also hardcodes "lib".
+# Without this setting, cmake by default would use the OS library
+# directory, which might be "lib64" or something else, not "lib", on multiarch systems.
 define $(package)_set_vars :=
-$(package)_config_opts := -DBUILD_TESTING=OFF
-$(package)_config_opts += -DWITH_OPENSSL=OFF
-$(package)_config_opts += -DWITH_ZLIB=OFF
+  $(package)_config_opts := -DBUILD_TESTING=OFF
+  $(package)_config_opts += -DWITH_OPENSSL=OFF
+  $(package)_config_opts += -DWITH_ZLIB=OFF
+  $(package)_config_opts += -DCMAKE_INSTALL_LIBDIR=lib/
 endef
 
 define $(package)_config_cmds

--- a/depends/packages/libmultiprocess.mk
+++ b/depends/packages/libmultiprocess.mk
@@ -14,6 +14,7 @@ endif
 # directory, which might be "lib64" or something else, not "lib", on multiarch systems.
 define $(package)_set_vars :=
 $(package)_config_opts += -DCMAKE_INSTALL_LIBDIR=lib/
+$(package)_config_opts += -DCMAKE_POSITION_INDEPENDENT_CODE=ON
 ifneq ($(host),$(build))
 $(package)_config_opts := -DCAPNP_EXECUTABLE="$$(native_capnp_prefixbin)/capnp"
 $(package)_config_opts += -DCAPNPC_CXX_EXECUTABLE="$$(native_capnp_prefixbin)/capnpc-c++"

--- a/depends/packages/libmultiprocess.mk
+++ b/depends/packages/libmultiprocess.mk
@@ -8,7 +8,12 @@ ifneq ($(host),$(build))
 $(package)_dependencies += native_capnp
 endif
 
+# Hardcode library install path to "lib" to match the PKG_CONFIG_PATH
+# setting in depends/config.site.in, which also hardcodes "lib".
+# Without this setting, cmake by default would use the OS library
+# directory, which might be "lib64" or something else, not "lib", on multiarch systems.
 define $(package)_set_vars :=
+$(package)_config_opts += -DCMAKE_INSTALL_LIBDIR=lib/
 ifneq ($(host),$(build))
 $(package)_config_opts := -DCAPNP_EXECUTABLE="$$(native_capnp_prefixbin)/capnp"
 $(package)_config_opts += -DCAPNPC_CXX_EXECUTABLE="$$(native_capnp_prefixbin)/capnpc-c++"

